File: haskell/minigrep.hs
{-
  Minigrep in Haskell
-}

import System.Environment (getArgs)
import System.Exit (exitFailure)
import System.IO (hPutStrLn, stderr)

main :: IO ()
main = do
  args <- getArgs
  case args of
    [query, filename] -> grep query filename
    _ -> do
      hPutStrLn stderr "Usage: minigrep <query> <filename>"
      exitFailure

grep :: String -> FilePath -> IO ()
grep query filename = do
  contents <- readFile filename
  let
    -- matchingLines = filter (query `isSubsequenceOf`) (lines contents)
    numberedLines = zip [1 ..] (lines contents)
    matches = [(n, line) | (n, line) <- numberedLines, query `isSubsequenceOf` line]

  if null matches
    then putStrLn $ "No matches found for '" ++ query ++ "' in " ++ filename
    else mapM_ printMatch matches
 where
  printMatch :: (Int, String) -> IO ()
  printMatch (lineNum, line) =
    putStrLn $ show lineNum ++ ": " ++ line

isSubsequenceOf :: (Eq a) => [a] -> [a] -> Bool
isSubsequenceOf [] _ = True
isSubsequenceOf _ [] = False
isSubsequenceOf needle@(x : xs) (y : ys)
  | x == y = isSubsequenceOf xs ys
  | otherwise = isSubsequenceOf needle ys

File: typescript/minigrep.ts
/**
 * Minigrep in TypeScript
 */

import fs from "fs";

// Get args
const file = process.argv[3];
const pattern = process.argv[2];
if (!file || !pattern) {
  console.error("Usage: minigrep.ts <pattern> <file>");
  process.exit(1);
}

// Get file contents
const contents = fs.readFileSync(file, "utf8");
if (!contents) {
  console.error("File not found");
  process.exit(1);
}

// Print matches
const lines = contents.split("\n");
lines.forEach((line, i) => {
  if (line.includes(pattern)) {
    console.log(`${i}: ${line}`);
  }
});

const numMatches = lines.filter((line) => line.includes(pattern)).length;
if (numMatches === 0) {
  console.log("No matches found");
  process.exit(1);
} else if (numMatches === 1) {
  console.log("Found 1 match");
} else {
  console.log(`Found ${numMatches} matches`);
}
